src/tmo/mantiuk08/pfstmo_mantiuk08.1
.TH "pfstmo_mantiuk08" 1
.SH NAME
pfstmo_mantiuk08 \- Display adaptive tone mapping
.SH SYNOPSIS
\fBpfstmo_mantiuk08\fR [\fB--display-function\fR <\fIdf-spec\fR>] [\fB--display-size\fR=<\fIsize-spec\fR>]
[\fB--color-saturation\fR <\fIfloat\fR>] [\fB--contrast-enhancement\fR <\fIfloat\fR>]
[\fB--white-y\fR=<\fIfloat\fR>] [\fB--fps\fR=<\fIframes-per-second\fR>]
[\fB--output-tone-curve\fR=<\fIfile name\fR>] [\fB--verbose\fR] [\fB--help\fR]
.SH DESCRIPTION
This command applies the display adaptive tone mapping, which attempts
to preserve contrast of an input (HDR) image as close as possible
given the characteristic of an output display. Use this tone mapping
operator if you want to preserve original image appearance, or
slightly enhance contrast (-\fBe\fR option) while maintaining the
natural look of images. The operator can also compensate for ambient
light reflections on a screen, and for varying dynamic range and
brightness of a display. The operator is suitable for video sequences
as it prevents high-frequency changes in tone-curve between
consecutive frames, which would result in flickering. Note that the
temporal filtering is always active and there is no need to specify
an argument to switch it on.
.HP
.PD 0
More details can be found in:
.IP
Rafal Mantiuk, Scott Daly and Louis Kerofsky.
.IP
Display Adaptive Tone Mapping.
.IP
In: ACM Transactions on Graphics 27 (3), 2008.
.IP
http://www.mpi-inf.mpg.de/resources/hdr/datmo/
.PD
.PP
If you find this TMO useful in your research project, please cite the
paper above.
.HP
.PD 0
This operator also employs color correction mechanism from:
.IP
Radoslaw Mantiuk, Rafal Mantiuk, Anna Tomaszewska, Wolfgang Heidrich.
.IP
Color Correction for Tone Mapping.
.IP
In: Computer Graphics Forum (Proc. of EUROGRAPHICS'09), 28(2), 2009.
.IP
http://zgk.wi.ps.pl/color_correction/
.PD
.PP
The result of this TMO does not require gamma correction.
.SH OPTIONS
.TP
\fB--display-function\fR <\fIdf-spec\fR>, \fB-d\fR <\fIdf-spec\fR>
To
adapt tone-mapping to different displays, this operator must be
provided a display function. The display function describes how output
luminance of a display changes with pixel values. If no parameter is
given, the command assumes \fB-df\ pd=lcd\fR (see \fIPre-defined
display\fR below). There are several ways to specify the display
function:
.TP
\fIGamma-gain-black-ambient display model\fR
.IP
g=<float>:l=<float>:b=<float>:k=<float>:a=<float>[:n=<float>]
.IP
Gamma-gain-black-ambient model can approximate a range of displays and
is a compact way to specify a display function. It assumes that a display function
has the following form:
.IP 
L_d(I) = (l-b)*I^gamma + b + k/pi*a
.IP
The parameters are as follows:
.RS
.PD 0
.TP 5
\fBg\fR - 
gamma or exponent of a display function (default 2.2, usually from 1.8 to 2.8)
.TP 5
\fBl\fR -
peak luminance of a display in cd/m^2  (default 100, from 80 for CRTs to 500 or more for newer displays)
.TP 5
\fBb\fR -
black level, which is luminance of a black pixel when the display is on (default 1, usually from 0.3 to 1 cd/m^2)
.TP 5
\fBk\fR -
reflectivity of a screen (assuming that it is diffuse) (default 0.01, usually about 0.01 (1%) for LCD displays, more for CRTs)
.TP 5
\fBa\fR -
ambient illumination in lux. Typical values are:
.RS
.IP 50\ lux
Family living room (dim, \fBdefault\fR)
.IP 400\ lux
A brightly lit office
.IP 32000\ lux
Sunlight on an average day (min.)
.IP 100000\ lux
Sunlight on an average day (max.)
.RE
.RE
.PD
.TP
\fIPre-defined display\fR
.IP
\fBpd\fR=\fI<display_type>\fI
.IP
Use pre-defined display type. This options are for convenience only
and they do not mean to accurately model the response of a particular
display.  The following \fIdisplay type\fRs are recognized:
.RS
.TP
\fBlcd_office\fR (g=2.2, l=100, b=0.8, k=0.01, a=400 )
lcd set to "office" mode seen in bright environment
.PD 0
.TP
\fBlcd\fR        (g=2.2, l=200, b=0.8, k=0.01, a=60  )
typical lcd seen in dim environment (\fBdefault\fR)
.TP
\fBlcd_bright\fR (g=2.6, l=500, b=0.5, k=0.01, a=10  )
newer LCD TV seen in dark environment
.TP
\fBcrt\fR        (g=2.2, l=80,  b=1,   k=0.02, a=60  )
CRT monitor seen in dim environment
.PD
.RE
.IP
The parameters
in the parenthesis are the same as for the gamma-gain-black-ambient
model explained above.
.TP
\fILookup-table\fR
.IP
\fBlut\fR=\fI<file>\fI
.IP
This is the most accurate specification of the display response
function, but requires measuring it with a luminance meter. The lookup
table should account also for ambient light, so that it is recommended
to use the luminance meter that can measure screen luminance from a
distance, such as Minolta LS-100 (as opposed to those that use rubber
tube touching a display that eliminates the influence of ambient
light). The <file> must be a comma-separated text file in a format
(CSV) with two columns: first column represents pixel values (from 0.0
to 1.0) and the second physical luminance in cd/m^2. Both the pixel
value and the luminance should increase in each raw.
.TP
\fB--display-size\fR=<\fIsize-spec\fR>, \fB-s\fR=<\fIsize_spec\fR>
Specifies how large the image appears to a viewer and what is the
viewing distance.  If no parameter is given, \fB-s\ ppd=30\fR is
assumed. Since this tone-mapper is global, display size has moderate
effect on the resulting images and thus skipping this parameter should
not do much harm. There are two ways to specify image size:
.RS
.PD 0
.TP
\fBvres\fR=<lines>:\fBvd\fR=<screen_heights>[:\fBd\fR=<\fImeters\fR>]
.RS
.IP \fBvres\fR
- screen's vertical resolution in lines, for example 1024. 
.IP \fBvd\fR
- viewing distance specified as multiplies of screen height. For example if the display is seen from 0.5m and the height of its screen is 25cm, \fBvd\fR=2.
.IP \fBd\fR
- (optional) viewing distance in meters. This is to account for lower
eye's sensitivity for larger viewing distances (although the effect is
negligible). By default \fB-d\fR=0.5 is assumed.
.RE
.TP
\fBppd\fR=<\fIpixels_per_visual_degree\fR>[:\fBd\fR=<\fImeters\fR>]
.RS
.IP \fBppd\fR
- how many pixels spans one visual degree.
.IP \fBd\fR
- (optional) viewing distance in meters. This is to account for lower
eye's sensitivity for larger viewing distances (although the effect is
negligible). By default \fB-d\fR=0.5 is assumed.
.RE
.RE
.PD
.TP
\fB--color-saturation\fR <\fIfloat\fR>, \fB-c\fR <\fIfloat\fR>
Decrease or increase color saturation after tone mapping. Default
value \fB-c=1\fR attempts to preserve color appearance of the original
image. Use values >1 to increase and <1 to decrease color saturation.
.TP
\fB--contrast-enhancement\fR <\fIfloat\fR>, \fB-e\fR <\fIfloat\fR>
By default this tone-mapper attempts to preserve contrast of an input
image (\fB-e=1\fR). This parameter controls whether the contrast of an
input image should be enhanced before tone-mapping. For example
\fB-e=1.15\fR boosts contrast by 15%. Note that if a target display
does not offer sufficient dynamic range, contrast may be enhanced only
for selected tone-values (those that dominate in an image) or not
enhanced at all.
.TP
\fB--white-y\fR=<\fIfloat\fR>, \fB-y\fR=<\fIfloat\fR>
Tells the
tone-mapper what luminance level in the input image should be mapped
to the maximum luminance of a display. Since HDR images contain only
relative luminance information, tone-mapper does not know how bright
should be the scene. This option is meant to fix this problem by
providing tone-mapper with the information what luminance level in an
input image should be perceived as a diffuse white surface. Default is
\fInone\fR, which means that no such mapping will be enforced and
tone-mapper is free to find an optimal brightness for a given
image. This is a recommended setting for HDR images. Setting
\fB--white-y\fR could be necessary for dark scenes, which could be
made too bright by the tone-mapper. The value of this parameter can be
also passed in pfsstream as a tag \fIWHITE_Y\fR. pfstools 1.7 and
newer sets set this tag automatically for LDR images. The command line
option overrides the value of the pfstream tag.
.TP
\fB--fps\fR=<\fIframes-per-second\fR>, \fB-f\fR=<\fIframes-per-second\fR>
Set the frame rate of the input sequence. Default is 25. Currently
only 3 values are supported: 25, 30 and 60. This parameter controls
temporal filter that makes sure the resulting sequence is coherent in
time. This reduces the likelihood of a visible flicker.
.TP
\fB--output-tone-curve\fR=<\fIfile name\fR>, \fB-o\fR=<\fIfile name\fR>
Write tone-curves to a text file. This option is mainly
for debugging purposes, but can be used to visualize computed
tone-curves. The tone-curve data is stored in a comma separated text
file (CSV), consisting of three columns: frame number, log10 of input
luminance factor, log10 of the resulting display luminance, and the
pixel value (0-1).
.TP
\fB--verbose\fR, \fB-v\fR
Print additional information during program execution.
.TP
\fB--quiet\fR, \fB-q\fR
Do not display progress report.
.TP
\fB--help\fR, \fB-h\fR
Print list of commandline options.
.SH EXAMPLES
.TP
pfsin memorial.hdr | pfstmo_mantiuk08 -d pd=crt | pfsout memorial.png
.IP
Tone map memorial image for a CRT display and store the result in the PNG format.
.TP
pfsin memorial.hdr | pfstmo_mantiuk08 -d g=2.6:l=500:b=0.5:k=0.01:a=10 | pfsview
.IP
Tone map memorial image for a display that has a 2.2 gamma, the peak
luminance of 500 cd/m^2, the black level of 0.5 cd/m^2, the panel
reflectivity of 1% (0.01) and is seen under the illumination of 10
lux. 
.TP
pfsin bridge.jpg --linear | pfsclamp --min 0.007 | pfstmo_mantiuk08 -v | pfsview
.IP
Enhance the low-dynamic range image 'bridge' and view the
result. pfsclamp command reduces noise for low code values.
.HP
.PD 0
pfsin frame%05d.exr | pfstmo_mantiuk08 -d pd=lcd_bright --fps 30 | pfsout out_frame%04d.png
.PD
.IP
Tone-map video sequence at 30 frame-per-second frame rate. 
.TP
pfsin *.exr | pfstmo_mantiuk08 | pfsview
.IP
Tone-map and display *.exr HDR images in the current directory. 
.TP
pfsin *.exr | pfstmo_mantiuk06 | pfsgamma -g 0.8 | pfstmo_mantiuk08 | pfsview
.IP
It is possible to stack a TMO that sharpens images (pfstmo_mantiuk06) with the
contrast preserving TMO (pfstmo_mantiuk08) to get new interesting results.
.SH "SEE ALSO"
.BR pfsin (1)
.BR pfsout (1)
.BR pfsview (1)
.SH BUGS
Please report bugs and comments to the pfstools discussion group
(http://groups.google.com/group/pfstools).


NEWS
This file is not maintained, but required by automake. Visit pfstools
web page to find the latest news:

http://pfstools.sourceforge.net/



src/tmo/mantiuk08/pfstmo_mantiuk08.cpp
/**
 * @brief Display Adaptive TMO
 *
 * From:
 * Rafal Mantiuk, Scott Daly, Louis Kerofsky.
 * Display Adaptive Tone Mapping.
 * To appear in: ACM Transactions on Graphics (Proc. of SIGGRAPH'08) 27 (3)
 * http://www.mpi-inf.mpg.de/resources/hdr/datmo/
 *
 * This file is a part of PFSTMO package.
 * ---------------------------------------------------------------------- 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * ---------------------------------------------------------------------- 
 * 
 * @author Rafal Mantiuk, <mantiuk@gmail.com>
 *
 * $Id: pfstmo_mantiuk08.cpp,v 1.19 2013/12/28 14:00:54 rafm Exp $
 */

//#include <config.h>

#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <getopt.h>
#include <fcntl.h>
#include <iostream>
#include <memory>

#include <sys/time.h>

#include <pfs.h>

#include "pfstmo_config.h"

#include "display_adaptive_tmo.h"


#define PROG_NAME "pfstmo_mantiuk08"

class QuietException 
{
};


class Timing
{
  timeval t1;
public:
  Timing() {
    gettimeofday(&t1, NULL);
  }

  void report( const char *activity )
  {
    timeval t2;
    gettimeofday(&t2, NULL);
    unsigned int t = (t2.tv_sec - t1.tv_sec) * 1000000 + (t2.tv_usec - t1.tv_usec);

    fprintf( stderr, "Activity %s took %g seconds.\n", activity,  (float)t / 1000000.f );
  }
};

using namespace std;

const char *temp_file_2pass = "datmo_tone_curves.tmp";
bool read_tone_curve( FILE *fh, datmoToneCurve *tc, double **x_scale );

void printHelp()
{
  fprintf( stderr, PROG_NAME " (" PACKAGE_STRING ") : \n"
    "\t[--display-function <df-spec>] [--display-size=<size-spec>]\n"
    "\t[--color-saturation <float>] [--contrast-enhancement <float>]\n"
    "\t[--white-y=<float>]\n"
    "\t[--verbose] [--quiet] [--help]\n"
    "See man page for more information.\n" );
}


bool verbose = false;
bool quiet = false;

#define FRAME_NAME_MAX 30
char frame_name[FRAME_NAME_MAX+1];

int progress_report( int progress )
{
  if( !quiet ) {  
    fprintf( stderr, "\r'%s' completed %d%%", frame_name, progress );
    if( progress == 100 )
      fprintf( stderr, "\n" );    
  }
  return PFSTMO_CB_CONTINUE;
}



void tmo_mantiuk08(int argc, char * argv[])
{

  //--- default tone mapping parameters;
  float contrast_enhance_factor = 1.f;
  float saturation_factor = 1.f;
  float white_y = -2.f;
  float fps = 25;  
  int temporal_filter = 0;
  int itmax = 200;
  float tol = 1e-3;
  DisplayFunction *df = NULL;
  DisplaySize *ds = NULL;
  const char *output_tc = NULL;
  datmoVisualModel visual_model = vm_full;
  double scene_l_adapt = 1000;  

  //--- process command line args

  df = createDisplayFunctionFromArgs( argc, argv );
  if( df == NULL )
    df = new DisplayFunctionGGBA( "lcd" );

  ds = createDisplaySizeFromArgs( argc, argv );
  if( ds == NULL )
    ds = new DisplaySize( 30.f, 0.5f );
  
  static struct option cmdLineOptions[] = {
    { "help", no_argument, NULL, 'h' },
    { "verbose", no_argument, NULL, 'v' },
    { "contrast-enhancement", required_argument, NULL, 'e' },
    { "color-saturation", required_argument, NULL, 'c' },
    { "white-y", required_argument, NULL, 'y' },
    { "fps", required_argument, NULL, 'f' },
    { "output-tone-curve", required_argument, NULL, 'o' },
    { "visual-model", required_argument, NULL, 'm' },
    { "scene-y-adapt", required_argument, NULL, 'a' },
    { "quiet", no_argument, NULL, 'q' },    
    { NULL, 0, NULL, 0 }
  };

  int optionIndex = 0;
  while( 1 ) {
    int c = getopt_long (argc, argv, "vhe:c:y:t:o:qm:f:a:", cmdLineOptions, &optionIndex);
    if( c == -1 ) break;
    switch( c ) {
    case 'h':
      printHelp();
      throw QuietException();
    case 'v':
      verbose = true;
      break;
    case 'q':
      quiet = true;
      break;
    case 'a':
      if( !strcmp( optarg, "auto" ) )
        scene_l_adapt = -1;
      else {        
        scene_l_adapt = (float)strtod( optarg, NULL );
        if( scene_l_adapt <= 0.0f )
        throw pfs::Exception("incorrect scane adaptation luminance. The value must be greater than 0.");
      }      
      break;
    case 'e':
      contrast_enhance_factor = (float)strtod( optarg, NULL );
      if( contrast_enhance_factor <= 0.0f )
        throw pfs::Exception("incorrect contrast enhancement factor, accepted value must be a positive number");
      break;
    case 'c':
      saturation_factor = (float)strtod( optarg, NULL );
      if( saturation_factor < 0.0f || saturation_factor > 2.0f )
        throw pfs::Exception("incorrect saturation factor, accepted range is (0..2)");
      break;
    case 'm':
    {      
      char *saveptr;
      char *token;
      visual_model = vm_none;
      
      token = strtok_r( optarg, ",:", &saveptr );
      while( token != NULL ) {
        
        if( !strcmp( token, "none" ) ) {
          visual_model = vm_none;
        } else if( !strcmp( token, "full" ) ) {      
          visual_model = vm_full;
        } else if( !strcmp( token, "luminance_masking" ) ) {          
          visual_model |= vm_luminance_masking;
        } else if( !strcmp( token, "contrast_masking" ) ) {          
          visual_model |= vm_contrast_masking;
        } else if( !strcmp( token, "csf" ) ) {          
          visual_model |= vm_csf;
        } else 
          throw pfs::Exception("Unrecognized visual model");
        
        token = strtok_r( NULL, ",:", &saveptr );
      }      
    }
    
      break;
      
    case 'y':
      if( !strcmp( optarg, "none" ) )
        white_y = -1;
      else
        white_y = (float)strtod( optarg, NULL );
      if( white_y < 0.0f )
        throw pfs::Exception("incorrect white-y value. The value must be greater than 0");
      break;
    case 'f':
      fps = strtod( optarg, NULL );
      if( fps != 25 && fps != 30 && fps != 60 )
        throw pfs::Exception("Only 3 frame-per-seconds values are supported: 25, 30 and 60.");
      break;
    case 'o':
      output_tc = optarg;
      break;
    case '?':
      throw QuietException();
    case ':':
      throw QuietException();
    }
  }

  if( verbose ) {
    df->print( stderr );
    ds->print( stderr );
    fprintf( stderr, "Frames-per-second: %g\n", fps );    
    fprintf( stderr, "Contrast masking: %d\n", (bool)(visual_model & vm_contrast_masking) );
    fprintf( stderr, "Luminance masking: %d\n", (bool)(visual_model & vm_luminance_masking) );
    fprintf( stderr, "CSF: %d\n", (bool)(visual_model & vm_csf) );
    fprintf( stderr, "Scane adaptation luminance: %g (-1 means auto)\n", scene_l_adapt );  
  }

  Timing tm_entire;

  FILE *tc_FH = NULL;
  if( output_tc != NULL ) {
    tc_FH = fopen( output_tc, "w" );
    if( tc_FH == NULL )
      throw pfs::Exception("cannot open file for writing tone-curve.");
  }

  
  datmoTCFilter rc_filter( fps, log10(df->display(0)), log10(df->display(1)) );
  pfs::DOMIO pfsio;

  size_t frame_no = 0;
  while( true ) {
    pfs::Frame *frame = pfsio.readFrame( stdin );
    if( frame == NULL )
      break;

    pfs::Channel *inX, *inY, *inZ;
	
    frame->getXYZChannels(inX, inY, inZ);
    int cols = frame->getWidth();
    int rows = frame->getHeight();

    const char *file_name = frame->getTags()->getString( "FILE_NAME" );
    if( file_name == NULL )
      sprintf( frame_name, "frame #%d", (int)frame_no );
    else {
      int len = strlen( file_name );
      if( len > FRAME_NAME_MAX ) // In case file name is too long
        len = FRAME_NAME_MAX-3;
      strcpy( frame_name, "..." );
      strncpy( frame_name+3, file_name + strlen( file_name ) - len, len+1 );
    }
    

    {
      pfs::Array2DImpl R( cols, rows );
  
      pfs::transformColorSpace( pfs::CS_XYZ, inX, inY, inZ, pfs::CS_RGB, inX, &R, inZ );

      if( white_y == -2.f ) {       // If not overriden by command line options
        const char *white_y_str = frame->getTags()->getString( "WHITE_Y" );
        if( white_y_str != NULL ) {
          white_y = strtof( white_y_str, NULL );
          if( white_y == 0 ) {
            white_y = -1;
            fprintf( stderr, PROG_NAME ": warning - wrong WHITE_Y in the input image" );        
          }
        }
      }      
      if( verbose && frame_no == 0 ) {
        fprintf( stderr, "Luminance factor of the reference white: " );
        if( white_y < 0 )
          fprintf( stderr, "not specified\n" );
        else
          fprintf( stderr, "%g\n", white_y );      
      }

      const char *lum_data = frame->getTags()->getString("LUMINANCE");
      if( lum_data != NULL && !strcmp( lum_data, "DISPLAY" ) && frame_no == 0 )
        fprintf( stderr, PROG_NAME " warning: input image should be in linear (not gamma corrected) luminance factor units. Use '--linear' option with pfsin* commands.\n" );

      Timing tm_cond_dens;
        std::auto_ptr<datmoConditionalDensity> C = datmo_compute_conditional_density( cols, rows, inY->getRawData(), progress_report );
        if( C.get() == NULL )
          throw pfs::Exception("failed to analyse the image");
	tm_cond_dens.report( "Conditional density" );

	Timing tm_comp_tone_curve;
        int res;
        datmoToneCurve *tc = rc_filter.getToneCurvePtr();        
        res = datmo_compute_tone_curve( tc, C.get(), df, ds, contrast_enhance_factor, white_y, progress_report, visual_model, scene_l_adapt );
        if( res != PFSTMO_OK )
          throw pfs::Exception( "failed to compute a tone-curve" );    

        datmoToneCurve *tc_filt = rc_filter.filterToneCurve();
	tm_comp_tone_curve.report( "Computing a tone-cuve" );
        

        {

	  Timing tm_tonecurve;
        int res;
        res = datmo_apply_tone_curve_cc( inX->getRawData(), R.getRawData(), inZ->getRawData(), cols, rows,
          inX->getRawData(), R.getRawData(), inZ->getRawData(), inY->getRawData(), tc_filt,
          df, saturation_factor );
        if( res != PFSTMO_OK )
          throw pfs::Exception( "failed to tone-map an image" );

	tm_tonecurve.report( "Apply tone-curve" );

        if( tc_FH != NULL ) {
          for( int i=0; i < tc_filt->size; i++ )
            fprintf( tc_FH, "%d,%g,%g,%g\n", (int)frame_no, tc_filt->x_i[i], tc_filt->y_i[i], df->inv_display( (float)pow( 10, tc_filt->y_i[i] ) ) );
        }
    
//    int res;
//    res = datmo_tonemap( inX->getRawData(), R.getRawData(), inZ->getRawData(), cols, rows,
//      inX->getRawData(), R.getRawData(), inZ->getRawData(), inY->getRawData(),
//    df, ds, contrast_enhance_factor, saturation_factor, white_y, progress_report );
        
        progress_report( 100 );        

        pfs::transformColorSpace( pfs::CS_RGB, inX, &R, inZ, pfs::CS_XYZ, inX, inY, inZ );
        frame->getTags()->setString("LUMINANCE", "DISPLAY");
  
        pfsio.writeFrame( frame, stdout );
      }
        
      pfsio.freeFrame(frame);
    }
    

    frame_no++;

  }

      
    if( tc_FH != NULL )
      fclose( tc_FH );
    
  delete df;
  delete ds;

  tm_entire.report( "Entire operation" );
}

bool read_tone_curve( FILE *fh, datmoToneCurve *tc, double **x_scale )
{
  int size, frame_no, read;
  read = fscanf( fh, "%d,%d\n", &frame_no, &size );
  if( read != 2 )
    return false;
  
  if( *x_scale == NULL )
    *x_scale = new double[size];

  tc->init( size, *x_scale );
  
  for( int i=0; i < size; i++ ) {
    float x, y;
    read = fscanf( fh, "%f,%f\n", &x, &y );
    if( read != 2 )
      throw pfs::Exception( "missing data in the 2-pass file"  );
    (*x_scale)[i] = x;
    tc->y_i[i] = y;
  }
  return true;
}



int main( int argc, char* argv[] )
{
  try {
    tmo_mantiuk08( argc, argv );
  }
  catch( pfs::Exception ex ) {
    fprintf( stderr, PROG_NAME " error: %s\n", ex.getMessage() );
    return EXIT_FAILURE;
  }        
  catch( QuietException  ex ) {
    return EXIT_FAILURE;
  }        
  return EXIT_SUCCESS;
}

